% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/analyse.R
\name{superFreq}
\alias{superFreq}
\title{Wrapper to run default superFreq analysis}
\usage{
superFreq(metaDataFile, captureRegions = "", normalDirectory, Rdirectory,
  plotDirectory, reference, genome = "hg19", BQoffset = 33, cpus = 3,
  outputToTerminalAsWell = T, forceRedo = forceRedoNothing(),
  normalCoverageDirectory = "", systematicVariance = 0.02,
  maxCov = 150, cloneDistanceCut = -qnorm(0.01), dbSNPdirectory = "",
  cosmicDirectory = "", resourceDirectory = "superFreqResources",
  mode = "exome", splitRun = T, participants = "all",
  manualStoryMerge = F, vepCall = "vep", correctReferenceBias = T,
  filterOffTarget = T, rareGermline = T, exacPopulation = "all",
  cosmicSalvageRate = 0.001, annotationMethod = "VariantAnnotation",
  ploidyPriors = NULL)
}
\arguments{
\item{metaDataFile}{Character. A path to a tab separated files with headers:
BAM: path (absolute or relative from metaData file) to bam file of sample.
     An index file is required as well.\cr
VCF: path (absolute or relative from metaData file) to vcf file of sample.
     This should include both somatic and germline variants. The variants
     will undergo further filtering, so it is not a problem if the .vcf
     includes false positives, although large number of entries in the
     .vcf increases run time.\cr
NAME: unique identifier of the sample. The names will be converted to
      standard R names, which involves replacing separators (such as space,
      dash or underscore) with dots.\cr
INDIVIDUAL: unique identifier of the indiviudal the sample comes from.
            This information is used pair up matched normal samples for
            somatic mutation filtering, and for identifying germline
            heterozygous SNPs. This column is also used to group samples
            that should be compared to each other: all samples from the
            same individual are analysed together, and mutations are
            tracked between the samples.\cr
NORMAL: Should be YES if the sample is normal, NO otherwise. A normal
        sample is assumed to have no somatic mutations. In practice, a
        tumor burden below a few percent works effectively as a normal,
        but tumor burdens above 5% in a sample marked as normal can cause
        severely reduced sensitivity in cancer samples from the same
        indiviudal.\cr
TIMEPOINT: Mostly used as label in plots together with the sample. Can
           be Diagnosis, Relapse, resistant or similar labels. Does not
           influence the analyss otherwise.}

\item{captureRegions}{Character. A path to a bed file of the unpadded capture regions.}

\item{normalDirectory}{Character. A path to a directory containing reference normals.
The directory should contain a subdirectory called "bam" with (links to)
.bam and index files for the reference normal
samples. These samples can, but dont have to, be matched normals
of the analysed samples. They need to be from the same capture as the
analysed samples, and preferably sequenced in the same lab. At least
two are required, but more are better (but slower). ~10 is a good number.
Another subdirectory called R will be created next to the bam directory.}

\item{Rdirectory}{Character. A path to the directory where data will be stored.
The directory will be created if needed, but the parent directory must exist.}

\item{plotDirectory}{Character. A path to the directory where plots and output is placed.
The directory will be created if needed, but the parent directory must exist.}

\item{reference}{Character. Path to the indexed fasta file the bams were aligned to.}

\item{genome}{Character. The reference genome the samples are aligned to, such as "hg19", "hg38"
or "mm10". Defaults to "hg19".}

\item{BQoffset}{Integer. The offset of the base quality encoding. Defaults to 33.}

\item{cpus}{Integer. Maximum number of parallel threads. Often threaded over chromsomes,
so numbers like 4, 6, 8, 12, 24 are good for human genomes. Default 3.}

\item{outputToTerminalAsWell}{Boolean. If the rather verbose log output should be sent to
terminal as well as to the logfile (in the Rdirectory).
Default TRUE.}

\item{forceRedo}{named list of Booleans. Controls which step are retrieved from previously
                 saved data, and which are redone even if previous results are saved.
                 If a step is forced to be redone, depending downstream steps are also
                 redone. The named entries in the list should be:\cr

                 forceRedoCount: The read counting over capture regions for the samples.\cr
                 forceRedoNormalCount: The read counting over capture regions for the
                                       set of reference normals.\cr
                 forceRedoFit: The differential coverage analysis.\cr
                 forceRedoVolcanoes: Volcano plots of the differential coverage.\cr
                 forceRedoDifferentRegions: Top table output of gained/lost genes/exons.\cr
                 forceRedoSNPs: Deprecated, usually not called.\cr
                 forceRedoVariants: Import and QC of the variants pointed to in the VCF. \cr
                 forceRedoNormalSNPs: Deprecated, usually not called.\cr
                 forceRedoNormalVariants: Import and QC of the variants in the set of
                                          reference normals.\cr
                 forceRedoMatchFlag: QC of variants using the set of reference normals.
                                     Somatic SNV are also separated from germline.\cr
                 forceRedoScatters: Scatter plots of the SNV frequencies between samples.\cr
                 forceRedoOutputSomatic: Top tables of somatic SNVs by sample.\cr
                 forceRedoNewVariants: Top table of new somatic SNVs compared to other
                                       samples.\cr
                 forceRedoSNPprogression: Heatmap of SNV frequencies over multiple samples.\cr
                 forceRedoCNV: Copy number calling.\cr
                 forceRedoCNVplots: Copy number plots.\cr
                 forceRedoSummary: Summary plot of SNVs and CNVs over all samples.\cr
                 forceRedoStories: Clonal tracking of somatic mutations.\cr
                 forceRedoRiver: Plots of the clonal evolution.\cr
                 forceRedoVEP: VEP call (Variant Effect Predictor) and COSMIC linking.
                 forceRedoSignature: The mutation signature analysis.

                 The default value is forceRedoNothing(), setting all entries to FALSE.
                 To redo everything, set to forceRedoEverything().}

\item{systematicVariance}{Numeric. A measure of the expected systematic variance in coverage
log fold change. Larger value decreases false CNV calls, smaller
values increase sensitivity. Defaults to 0.03.}

\item{maxCov}{Integer. The coverage at which systematic variance of SNV frequencies are equal
to the Poissonian variance. Larger values increase sensitivity in CNV
calls, smaller values decrease false positives. Default 150.}

\item{dbSNPdirectory}{Character. Deprecated. Use resourceDirectory instead.}

\item{cosmicDirectory}{Character. Deprecated. Use resourceDirectory instead.}

\item{resourceDirectory}{Character. The location of the directory where superFreq resources
are located. Such as dbSNP, COSMIC, annotation.
If the directory doesn't exist, it will be created and the
data will be downloaded. Defaults to superFreqResources (in the directory
where R is run).}

\item{mode}{Character. The mode to run in. The default 'Exome' is almost always used. If running on RNA,
switch to "RNA", which seems to work decently, but beware of limitations of the data.}

\item{splitRun}{logical. If the superFreq run should be split by individual. Runs each individual separately,
which decrease memory footprint. Default TRUE.}

\item{participants}{character. When run in split mode, this allows you to run on a subset of individuals.
Default is "all" which runs on all individuals.}

\item{manualStoryMerge}{logical. A flag which gives you some manual control over the merging of mutations into clones.
Requires babysitting and not recommended. Do not press this button. Default FALSE.}

\item{vepCall}{character. The call for the variant effect predictor on the command line. Default "vep",
otherwise try the path to variant_effect_predictor.pl.}

\item{correctReferenceBias}{logical. Corrects for (mainly alignment-caused) bias towards the reference allele in VAF
calculations. Default TRUE.}

\item{filterOffTarget}{logical. Ignores any point variants more than 300bp outside the capture regions. Default TRUE.}

\item{rareGermline}{logical. Display rare germline variants. This is a by-product of the clonal tracking and can yield
important information about the sample, but due to ethics this is not always desired.
This options allows a user to run superFreq while avoiding germline information. The user
should keep in mind that superFreq, like most open source software, comes with no warranty and
there is always a risk of bugs producing unintended behaviour. Also note that for individuals without matched normal it can be difficult to separate cancer variants present in most cells in all samples from germline variants. So rare germline variants may be misidentified as somatic variants without a matched normal.}

\item{exacPopulation}{character. Which population to use for the exac data. This influences annotation, selection of
heterozygous SNPs for variant calling and, in case of no matched normal, identification of
somatic variants. Default 'all' uses the total allele frequency across all of (non-TCGA) ExAC.
Other available options are 'AFR' (Africa), 'AMR' (America), 'EAS' (East Asia),
'FIN' (Finland), 'NFE' (Non-Finland Europe), 'OTH' (Other) and 'SAS' (South Asia).}

\item{annotationMethod}{character. Which variant annotator to use. 'VEP' is the legacy method and requires a system
installation of VEP. VariantAnnotation is contained in R and is significantly faster.
Default 'VariantAnnotation'.}

\item{ploidyPriors}{named numeric vector. Prior belief or knowledge that strongly biases the analysis to pick a ploidy close to the supplied prior. Names of the vector must match the samples you want to affect, but doesnt have to include all analysed samples. Entries not matching sample names are ignored. Note that this is sample-wide ploidy, including any normal contamination, so for example a 0.5 purity samples where the cancer has a AABB genotype genomewide will have a sample-wide ploidy of 3. If this is for a fix-the-analysis second run, it might be informative to look at the fitness curves in plotDirectory, diagnostics, ploidy of the first run to find the secondary local minimum that might be the ploidy you want to be used. Default NULL is a flat prior on ploidy, although the prior preference for less exotic copy numbers implicetely biases the analysis towards ploidies close to 2.}
}
\description{
Wrapper to run default superFreq analysis
}
\details{
This function runs a full SNV, CNV and clonality analysis of the input exome data. Output it sent to the plotDirectory, where diagnostics as well as analysis results are placed. At 12 cpus, the pipeline typically runs overnight on an individual with not too many samples. Large numbers of samples or somatic mutations can increase runtime further. Read more on the manual on https://github.com/ChristofferFlensburg/superFreq/blob/master/manual.pdf about how to set up the run and interpret the output.
}
\examples{
\dontrun{
#minimal example to get it running, assuming data fits the defaults.
#you really want to skim through the other settings before running though.
metaDataFile = 'metaData.tsv'
captureRegions='captureRegions.bed'
normalDirectory = '../referenceNormals'
reference = '../reference/hg19.fa'
Rdirectory = 'R'
plotDirectory='plots'

data = superFreq(metaDataFile, captureRegions, normalDirectory,
                 Rdirectory, plotDirectory, reference)
}
}
